Gothic-Unity-Project/Gothic-Unity
# NPC culling - v1

I implemented the first iteration of NPC culling. 🥳 

## To test
* Alternate tests between starting a new game and loading a save game. Both should work.
* NPCs should become invisible after 50m of distance to player.
* No Frustum Culling or Occlusion Culling should disable NPCs. Only distance.

### It works, if
1. If an NPC walks out of your 50m range, then they disappear
2. When reappearing, they will spawn at their initial Waypoint again (this is normal G1 behavior. I tested it with G1's Marvin mode)
3. @see special Diego test below

### Special test for Diego
   * Diego is a wonderful test subject. When a new game starts, he has a different routine than after he talked to us for the first time.
   * Please create a save game in original G1 where he started walking.
   * Then load the save into our game and see him walking.
   * When he reaches 50m, he will disappear.
   * His next appearance will be, when you walk to Old Camp.
   * This is expected behaviour and works the same in original G1


## Technical updates:
* Culling is an optional flag. If we disable culling, NPCs will become disabled until the world is fully loaded (intentionally)
* Once the solution is tested and approved, there are a few missing pieces left, I track them here: #77 
* I also optimized a few other topics mostly related to culling and loading. Please check commit comments if you're interested.
* There will be a lot of warnings and some Ai errors when you start a new game or run it for a while. The warnings are related to unimplemented Daedalus Externals, because we now load all NPCs/Monsters in the world with their untested logic. And I will look at the AI exceptions later.
* I started the game about 30 times with New Game or Load Game. I also tested with HVR and XRIT. Most tests were done in the Starting area. Feel free to test other parts of the world.


### Further bug fixes/features within this branch
* NPCs are now loaded from a save game or new game's Wld_InsertNpc() calls. (Hint: Only the nearby NPCs are loaded so far (different list in Save Game state). Not the ones from over the world. I will work on that later.)
* Loading of NPCs is now done async like VOB and World loading.
* I optimized HVR player teleportation to start point. Now it is directly when the Player is created from Prefab. Not a second later. (Fix was needed for Culling to work properly)
* I changed loading of NPCs to STARTUP_WORLDNAME() instead of STARTUP_OLDCAMP(). --> Now whenever we load a world, it will load the NPCs/Monsters from that whole world/scene, not just OC. (Nevertheless, I didn't test the world switching so far.)
*
